raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -435,16 +438,40 @@ private DescribeQuorumResponseData.ReplicaState describeReplicaState(
 
     }
 
+    /**
+     * Clear observer states that have not been active for a while and are not the leader.
+     */
     private void clearInactiveObservers(final long currentTimeMs) {
         observerStates.entrySet().removeIf(integerReplicaStateEntry ->
-            currentTimeMs - integerReplicaStateEntry.getValue().lastFetchTimestamp >= OBSERVER_SESSION_TIMEOUT_MS
+            currentTimeMs - integerReplicaStateEntry.getValue().lastFetchTimestamp >= OBSERVER_SESSION_TIMEOUT_MS &&
+                integerReplicaStateEntry.getKey() != localId
         );
     }
 
     private boolean isVoter(int remoteNodeId) {
         return voterStates.containsKey(remoteNodeId);
     }
 
+    private void updateVoterAndObserverStates(Set<Integer> lastVoterSet) {
```

Comment:
I didn't see we have any unit tests for this. Could we add some to verify the observerState/voterState are changed as expected?

Reply:
I think we can test this indirectly with making sure observers cannot influence HW (e.g. if we have two voters, two observers, all observers being up-to-date with leader will _not_ cause HW to increase) until they are added back to the voterSet. This ensures nodes are correctly removed/added from the sets

Reply:
Hm, actually seems we can test this pretty directly with `describeQuorum`. I'll add another test.